### ISSUE TITLE
Implement support for global mute per tab

### ISSUE THREAD

**Issue by ferjm** [servo/servo#22359]
It would be nice to have a way to implement the global mute per tab feature.

Right now, we have an independent [Player](https://github.com/servo/media/blob/master/player/src/lib.rs#L46) instance per `HTMLMediaElement`. One way we could do this is to keep a match of all these instances per top level browsing context (@jdm @asajeffrey  do we have the concept of principal in Servo like we have in Gecko?) and expose a way to mute/unmute all of them. Most of this work will likely happen in [servo/media](https://github.com/servo/media).

Muting a tab does not affect the `muted` value of any of the `HTMLMediaElement`s contained by the tab on Firefox or Chrome.

**Comment by ferjm** [comment]
This should not only affect audio associated to `HTMLMediaElement`s. We should also keep track of WebAudio `AudioContext` instances.

**Comment by ferjm** [comment]
https://github.com/servo/media/blob/master/audio/src/context.rs#L105

**Comment by Tommytrg** [comment]
Hi, I would like to work on this issue. How could I start?

**Comment by cbrewster** [comment]
@ferjm you can store per-top level browsing context info in the `Browser` struct in the constellation:

https://github.com/servo/servo/blob/16e4eb69644b74d120398cc2b3ac0d9a94075ae5/components/constellation/constellation.rs#L171-L181

**Comment by ferjm** [comment]
Thanks for your interest in working on this @Tommytrg.

I think we could do something like:

- Make ServoMedia take a generic parameter [here](https://github.com/servo/media/blob/ce1cc2d18f95776c66db72dfb59685fe36f0de4d/servo-media/src/lib.rs#L13) to get the type of identifiers the client uses for their media contexts (top-level browsing contexts in Servo's case). For example
```rust
pub struct ServoMedia<MediaContextId>
```
- Create a `Muteable` trait with a single `mute` method.
- Make [Player](https://github.com/servo/media/blob/ce1cc2d18f95776c66db72dfb59685fe36f0de4d/player/src/lib.rs#L46) and [AudioContext](https://github.com/servo/media/blob/ce1cc2d18f95776c66db72dfb59685fe36f0de4d/audio/src/context.rs#L105) implement this trait.
- Add an new member (named i.e.`mutables`) to the `ServoMedia` struct to store a match between these context ID's and `Rc<Muteable>`s .
- Add an extra `ClientContextId` argument to [create_audio_context](https://github.com/servo/media/blob/ce1cc2d18f95776c66db72dfb59685fe36f0de4d/servo-media/src/lib.rs#L71) and [create_player](https://github.com/servo/media/blob/ce1cc2d18f95776c66db72dfb59685fe36f0de4d/servo-media/src/lib.rs#L75). You'll likely need to make these methods return `Rc<AudioContext<Backend>>` and `Rc<Player<Error=Error>>` respectively.
- In the implementation of these two methods, store the match between the `ClientContextId` and the Rc<Muteable> in `self.muteables`.
- Expose a new method (named i.e. `mute_media_context`) in [ServoMedia](https://github.com/servo/media/blob/ce1cc2d18f95776c66db72dfb59685fe36f0de4d/servo-media/src/lib.rs#L53) to go through the list of stored `Mutable`s and call their `mute` methods.

Don't hesitate to ask if you have any question.

**Comment by ferjm** [comment]
@Tommytrg how are things going with this issue? Do not hesitate to ask if you need any further help.

**Comment by mandreyel** [comment]
@Tommytrg Hey! Are you still working on this issue? Please let me know as I'd also be interested in working on it.

**Comment by Tommytrg** [comment]
@mandreyel I'm still working on it. Sorry for the delay...

**Comment by ferjm** [comment]
Ugh, I am afraid that my recommendations from https://github.com/servo/servo/issues/22359#issuecomment-445038353 lead you to a missing Rust feature: [generic global variables](https://internals.rust-lang.org/t/how-about-generic-global-variables/8351). Sorry about that :(.

I guess we could get a tuple of numbers as IDs, as that's what Servo uses in the end to identify tabs ([BrowsingContextId](https://github.com/servo/servo/blob/16e4eb69644b74d120398cc2b3ac0d9a94075ae5/components/msg/constellation_msg.rs#L153) contains a [PipelineNameSpaceId](https://github.com/servo/servo/blob/16e4eb69644b74d120398cc2b3ac0d9a94075ae5/components/msg/constellation_msg.rs#L84) which is an `u32` and a [BrowsingContextIndex](https://github.com/servo/servo/blob/16e4eb69644b74d120398cc2b3ac0d9a94075ae5/components/msg/constellation_msg.rs#L147) which is a `NonZeroU32`). But I'd really prefer to make this generic. @Manishearth might have ideas for a better approach to implement this.

> By the other hand, what it's the type of the ClientContextId that I have to add as an argument? I think I don't really understand what are those `Id's.

That would be [BrowsingContextId](https://github.com/servo/servo/blob/16e4eb69644b74d120398cc2b3ac0d9a94075ae5/components/msg/constellation_msg.rs#L153) in Servo's case.

These IDs are the tab (top level browsing context) identifiers. We need a way to mute per tab, so we need to keep track of which `Players` and `AudioContext` instances belong to which tab, and to do that we need to have a match between these instances and the tabs identifiers.

**Comment by Hyperion101010** [comment]
ferjm i think Tommytrg isn't working on it ,so much of delay !

**Comment by ferjm** [comment]
@Tommytrg how are things going?

I think we can start with the tuple of numbers for the IDs as suggested before to unblock your work, and we can revisit this later to try to make it more generic.

**Comment by Tommytrg** [comment]
I think I have delayed so much and I want to stop blocking this issue. I would try to solve an easier one in the future. Thank you very much for your help and sorry for waste your time.

**Comment by 192-sean** [comment]
Hey I'd love to work on this issue.

**Comment by KiChjang** [comment]
@192-sean Go for it!

**Comment by ferjm** [comment]
@Tommytrg no worries and thanks for your efforts!

@192-sean thanks! Let me know if you need any help with this.

**Comment by 192-sean** [comment]
For the ServoMedia new member muteables, are you thinking something like a vector of top level browser id and Rc<Muteable>.

**Comment by ferjm** [comment]
A `Vec` would work, but it'd probably be better to use a `HashMap` instead, so get operations are O(1).

**Comment by 192-sean** [comment]
muteables: HashMap<clientContextId, Rc<Muteable>>; 
I am using this data structure but having two problems. 
First clientContextId has multiple values so maybe Rc<Vec<Muteable>>.
Second I can't call method mute() on a reference so should I use RefCell?

**Comment by ferjm** [comment]
> First clientContextId has multiple values so maybe Rc<Vec<Muteable>>.

Yes, we need a collection here, but since we don't care about the order, a `HashSet` might be a better fit in this case. Also we probably need a collection of `Rc<Muteable>`s, so I'd try with a `HashSet<Rc<Muteable>>`.

>  Second I can't call method mute() on a reference so should I use RefCell?

You should be able to call `mute()` on a trait object like `Rc<Muteable>`. And I don't think you need interior mutability here.

Feel free to open a PR with your work in progress patch. It's easier to give feedback having the actual code :)

**Comment by ferjm** [comment]
@192-sean how are things going here? Are you still working on this issue?

**Comment by 192-sean** [comment]
Yes I will submit a pr shortly.

**Comment by ferjm** [comment]
@192-sean cool. Let me know if you need any help.

**Comment by ferjm** [comment]
@192-sean ping :) Don't hesitate to ask if you need any help moving this issue forward, please.

**Comment by ferjm** [comment]
Unassigning due to lack of activity. Feel free to assign it back to you if you restart the work on this.

**Comment by oneturkmen** [comment]
hey @ferjm, I want to take this is my first "less-easy" task :) thanks for the [thorough description](https://github.com/servo/servo/issues/22359#issuecomment-445038353). I shall be back if I have any questions.

**Comment by oneturkmen** [comment]
hey @ferjm , I am currently working on some other tasks on servo and I feel like I gotta more through the rust book to fully understand what's going on. Could you un-assign me for now (since there may be someone interested meanwhile)? If I am done soon and there is no one on this issue, I will come back :100:

**Comment by ferjm** [comment]
@oneturkmen no worries. Feel free to get back to the issue if you find the time to work on it. Thanks!

**Comment by jdm** [comment]
@ferjm Has this been completed?

**Comment by ferjm** [comment]
Yes. https://github.com/servo/media/pull/271 implemented the necessary bits and it's already integrated in Servo. There are some remaining tweaks to be done in follow-ups like https://github.com/servo/media/pull/279 but this one can be closed now.
